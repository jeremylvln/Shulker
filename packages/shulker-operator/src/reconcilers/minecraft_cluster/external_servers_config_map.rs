use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;
use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

use super::MinecraftClusterReconciler;

pub struct ExternalServersConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for ExternalServersConfigMapBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = ConfigMap;
    type Context = ();

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-external-servers", cluster.name_any())
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    fn is_needed(&self, cluster: &Self::OwnerType) -> bool {
        cluster
            .spec
            .external_servers
            .as_ref()
            .map_or(false, |list| !list.is_empty())
    }

    async fn build(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
        _existing_config_map: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "external-servers".to_string(),
                    "proxy".to_string(),
                )),
                ..ObjectMeta::default()
            },
            data: Some(BTreeMap::from([(
                "external-servers.yaml".to_string(),
                ExternalServersConfigMapBuilder::get_content_from_server_list(cluster),
            )])),
            ..ConfigMap::default()
        };

        Ok(config_map)
    }
}

impl ExternalServersConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ExternalServersConfigMapBuilder { client }
    }

    fn get_content_from_server_list(cluster: &MinecraftCluster) -> String {
        serde_yaml::to_string(cluster.spec.external_servers.as_ref().unwrap()).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

    use crate::reconcilers::minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER};

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::ExternalServersConfigMapBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-external-servers");
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ExternalServersConfigMapBuilder::new(client);
        let name = super::ExternalServersConfigMapBuilder::name(&TEST_CLUSTER);

        // W
        let config_map = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }
}
