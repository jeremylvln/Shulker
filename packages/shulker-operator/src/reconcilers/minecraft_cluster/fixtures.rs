use http::{Request, Response};
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::v1alpha1::minecraft_cluster::{
    MinecraftCluster, MinecraftClusterExternalServerSpec, MinecraftClusterSpec,
};

lazy_static! {
    pub static ref TEST_CLUSTER: MinecraftCluster = MinecraftCluster {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-cluster".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftClusterSpec {
            network_admins: None,
            redis: None,
            external_servers: Some(vec![MinecraftClusterExternalServerSpec {
                name: "my-external-server".to_string(),
                address: "127.0.0.1:25565".to_string(),
                tags: vec!["game".to_string()]
            }])
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
