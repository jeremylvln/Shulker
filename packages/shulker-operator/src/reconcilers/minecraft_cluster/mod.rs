use std::{collections::BTreeMap, sync::Arc, time::Duration};

use external_servers_config_map::ExternalServersConfigMapBuilder;
use futures::StreamExt;
use k8s_openapi::api::{
    apps::v1::StatefulSet,
    core::v1::{ConfigMap, Secret, Service, ServiceAccount},
    rbac::v1::{Role, RoleBinding},
};
use kube::{
    api::ListParams,
    runtime::{
        controller::Action,
        finalizer::{finalizer, Event as Finalizer},
        watcher::Config,
        Controller,
    },
    Api, Client, ResourceExt,
};
use shulker_kube_utils::reconcilers::builder::reconcile_builder;
use tracing::*;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;

use crate::reconcilers::ReconcilerError;

use self::{
    forwarding_secret::ForwardingSecretBuilder, headless_service::HeadlessServiceBuilder,
    minecraft_server_role::MinecraftServerRoleBuilder,
    minecraft_server_role_binding::MinecraftServerRoleBindingBuilder,
    minecraft_server_service_account::MinecraftServerServiceAccountBuilder,
    proxy_role::ProxyRoleBuilder, proxy_role_binding::ProxyRoleBindingBuilder,
    proxy_service_account::ProxyServiceAccountBuilder, redis_service::RedisServiceBuilder,
    redis_stateful_set::RedisStatefulSetBuilder,
};

use super::Result;

pub mod external_servers_config_map;
mod forwarding_secret;
mod headless_service;
mod minecraft_server_role;
mod minecraft_server_role_binding;
mod minecraft_server_service_account;
mod proxy_role;
mod proxy_role_binding;
mod proxy_service_account;
pub mod redis_service;
mod redis_stateful_set;

#[cfg(test)]
pub mod fixtures;

static FINALIZER: &str = "minecraftclusters.shulkermc.io";

struct MinecraftClusterReconciler {
    client: kube::Client,

    // Builders
    forwarding_secret_builder: ForwardingSecretBuilder,
    headless_service_builder: HeadlessServiceBuilder,
    proxy_service_account_builder: ProxyServiceAccountBuilder,
    proxy_role_builder: ProxyRoleBuilder,
    proxy_role_binding_builder: ProxyRoleBindingBuilder,
    minecraft_server_service_account_builder: MinecraftServerServiceAccountBuilder,
    minecraft_server_role_builder: MinecraftServerRoleBuilder,
    minecraft_server_role_binding_builder: MinecraftServerRoleBindingBuilder,
    redis_service_builder: RedisServiceBuilder,
    redis_stateful_set_builder: RedisStatefulSetBuilder,
    external_servers_config_map_builder: ExternalServersConfigMapBuilder,
}

impl MinecraftClusterReconciler {
    async fn reconcile(
        &self,
        _api: Api<MinecraftCluster>,
        cluster: Arc<MinecraftCluster>,
    ) -> Result<Action> {
        reconcile_builder(&self.forwarding_secret_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.headless_service_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.proxy_service_account_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.proxy_role_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.proxy_role_binding_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(
            &self.minecraft_server_service_account_builder,
            cluster.as_ref(),
            None,
        )
        .await
        .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.minecraft_server_role_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(
            &self.minecraft_server_role_binding_builder,
            cluster.as_ref(),
            None,
        )
        .await
        .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.redis_service_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.redis_stateful_set_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(
            &self.external_servers_config_map_builder,
            cluster.as_ref(),
            None,
        )
        .await
        .map_err(ReconcilerError::BuilderError)?;

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, cluster: Arc<MinecraftCluster>) -> Result<Action> {
        info!(
            name = cluster.name_any(),
            namespace = cluster.namespace(),
            "cleaning up MinecraftCluster",
        );

        Ok(Action::await_change())
    }

    fn get_labels(
        cluster: &MinecraftCluster,
        name: String,
        component: String,
    ) -> BTreeMap<String, String> {
        BTreeMap::from([
            ("app.kubernetes.io/name".to_string(), name.clone()),
            (
                "app.kubernetes.io/instance".to_string(),
                format!("{}-{}", name, cluster.name_any()),
            ),
            ("app.kubernetes.io/component".to_string(), component),
            (
                "app.kubernetes.io/part-of".to_string(),
                format!("cluster-{}", cluster.name_any()),
            ),
            (
                "app.kubernetes.io/managed-by".to_string(),
                "shulker-operator".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                cluster.name_any(),
            ),
        ])
    }
}

#[instrument(skip(ctx, cluster))]
async fn reconcile(
    cluster: Arc<MinecraftCluster>,
    ctx: Arc<MinecraftClusterReconciler>,
) -> Result<Action> {
    let ns = cluster.namespace().unwrap();
    let clusters_api: Api<MinecraftCluster> = Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = cluster.name_any(),
        namespace = ns,
        "reconciling MinecraftCluster",
    );

    finalizer(&clusters_api, FINALIZER, cluster, |event| async {
        match event {
            Finalizer::Apply(cluster) => ctx.reconcile(clusters_api.clone(), cluster.clone()).await,
            Finalizer::Cleanup(cluster) => ctx.cleanup(cluster.clone()).await,
        }
    })
    .await
    .map_err(|e| ReconcilerError::FinalizerError(Box::new(e)))
}

fn error_policy(
    _cluster: Arc<MinecraftCluster>,
    error: &ReconcilerError,
    _ctx: Arc<MinecraftClusterReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&cluster, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client) {
    let clusters_api = Api::<MinecraftCluster>::all(client.clone());
    if let Err(e) = clusters_api.list(&ListParams::default().limit(1)).await {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = MinecraftClusterReconciler {
        client: client.clone(),
        forwarding_secret_builder: ForwardingSecretBuilder::new(client.clone()),
        headless_service_builder: HeadlessServiceBuilder::new(client.clone()),
        proxy_service_account_builder: ProxyServiceAccountBuilder::new(client.clone()),
        proxy_role_builder: ProxyRoleBuilder::new(client.clone()),
        proxy_role_binding_builder: ProxyRoleBindingBuilder::new(client.clone()),
        minecraft_server_service_account_builder: MinecraftServerServiceAccountBuilder::new(
            client.clone(),
        ),
        minecraft_server_role_builder: MinecraftServerRoleBuilder::new(client.clone()),
        minecraft_server_role_binding_builder: MinecraftServerRoleBindingBuilder::new(
            client.clone(),
        ),
        redis_service_builder: RedisServiceBuilder::new(client.clone()),
        redis_stateful_set_builder: RedisStatefulSetBuilder::new(client.clone()),

        external_servers_config_map_builder: ExternalServersConfigMapBuilder::new(client.clone()),
    };

    Controller::new(clusters_api, Config::default().any_semantic())
        .owns(
            Api::<ConfigMap>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Secret>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<ServiceAccount>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Role>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<RoleBinding>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Service>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<StatefulSet>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .shutdown_on_signal()
        .run(reconcile, error_policy, context.into())
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;
}
