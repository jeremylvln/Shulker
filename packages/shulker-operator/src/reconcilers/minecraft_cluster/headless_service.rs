use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::Service;
use k8s_openapi::api::core::v1::ServiceSpec;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;
use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

use super::MinecraftClusterReconciler;

pub struct HeadlessServiceBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for HeadlessServiceBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = Service;
    type Context = ();

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-cluster", cluster.name_any())
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    async fn build(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
        _existing_service: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let service = Service {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "service".to_string(),
                    "minecraft-server-headless".to_string(),
                )),
                ..ObjectMeta::default()
            },
            spec: Some(ServiceSpec {
                selector: Some(BTreeMap::from([(
                    "minecraftcluster.shulkermc.io/name".to_string(),
                    cluster.name_any(),
                )])),
                type_: None,
                cluster_ip: Some("None".to_string()),
                ports: Some(vec![]),
                ..ServiceSpec::default()
            }),
            ..Service::default()
        };

        Ok(service)
    }
}

impl HeadlessServiceBuilder {
    pub fn new(client: Client) -> Self {
        HeadlessServiceBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

    use crate::reconcilers::minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER};

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::HeadlessServiceBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-cluster");
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::HeadlessServiceBuilder::new(client);
        let name = super::HeadlessServiceBuilder::name(&TEST_CLUSTER);

        // W
        let service = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(service);
    }
}
