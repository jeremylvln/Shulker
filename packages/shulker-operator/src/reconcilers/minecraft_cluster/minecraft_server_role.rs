use k8s_openapi::api::rbac::v1::PolicyRule;
use k8s_openapi::api::rbac::v1::Role;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use super::MinecraftClusterReconciler;
use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;
use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

pub struct MinecraftServerRoleBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for MinecraftServerRoleBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = Role;
    type Context = ();

    fn name(cluster: &Self::OwnerType) -> String {
        format!("shulker:{}:server", cluster.name_any())
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    async fn build(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
        _existing_role: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let role = Role {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "role".to_string(),
                    "minecraft-server-rbac".to_string(),
                )),
                ..ObjectMeta::default()
            },
            rules: Some(vec![
                PolicyRule {
                    api_groups: Some(vec!["agones.dev".to_string()]),
                    resources: Some(vec!["gameservers".to_string()]),
                    verbs: vec![
                        "list".to_string(),
                        "watch".to_string(),
                        "update".to_string(),
                        "patch".to_string(),
                    ],
                    ..PolicyRule::default()
                },
                PolicyRule {
                    api_groups: Some(vec!["".to_string()]),
                    resources: Some(vec!["events".to_string()]),
                    verbs: vec!["create".to_string(), "patch".to_string()],
                    ..PolicyRule::default()
                },
            ]),
        };

        Ok(role)
    }
}

impl MinecraftServerRoleBuilder {
    pub fn new(client: Client) -> Self {
        MinecraftServerRoleBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

    use crate::reconcilers::minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER};

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::MinecraftServerRoleBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "shulker:my-cluster:server");
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBuilder::new(client);
        let name = super::MinecraftServerRoleBuilder::name(&TEST_CLUSTER);

        // W
        let role = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(role);
    }

    #[tokio::test]
    async fn build_can_create_events() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBuilder::new(client);
        let name = super::MinecraftServerRoleBuilder::name(&TEST_CLUSTER);

        // W
        let role = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        assert!(role.rules.as_ref().unwrap().iter().any(|rule| {
            rule.api_groups == Some(vec!["".to_string()])
                && rule.resources == Some(vec!["events".to_string()])
                && rule.verbs.contains(&"create".to_string())
        }));
    }

    #[tokio::test]
    async fn build_can_watch_gameservers() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBuilder::new(client);
        let name = super::MinecraftServerRoleBuilder::name(&TEST_CLUSTER);

        // W
        let role = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        assert!(role.rules.as_ref().unwrap().iter().any(|rule| {
            rule.api_groups == Some(vec!["agones.dev".to_string()])
                && rule.resources == Some(vec!["gameservers".to_string()])
                && rule.verbs.contains(&"list".to_string())
        }));
        assert!(role.rules.as_ref().unwrap().iter().any(|rule| {
            rule.api_groups == Some(vec!["agones.dev".to_string()])
                && rule.resources == Some(vec!["gameservers".to_string()])
                && rule.verbs.contains(&"watch".to_string())
        }));
    }

    #[tokio::test]
    async fn build_can_update_gameservers() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBuilder::new(client);
        let name = super::MinecraftServerRoleBuilder::name(&TEST_CLUSTER);

        // W
        let role = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        assert!(role.rules.as_ref().unwrap().iter().any(|rule| {
            rule.api_groups == Some(vec!["agones.dev".to_string()])
                && rule.resources == Some(vec!["gameservers".to_string()])
                && rule.verbs.contains(&"update".to_string())
        }));
    }
}
