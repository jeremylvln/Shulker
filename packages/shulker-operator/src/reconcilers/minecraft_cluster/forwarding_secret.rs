use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::Secret;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;
use rand::distributions::{Alphanumeric, DistString};

use crate::reconcilers::builder::ResourceBuilder;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;

use super::MinecraftClusterReconciler;

const SECRET_DATA_KEY: &str = "key";

pub struct ForwardingSecretBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ForwardingSecretBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = Secret;

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-forwarding-secret", cluster.name_any())
    }

    fn is_updatable() -> bool {
        false
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    fn is_needed(&self, _cluster: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let secret = Secret {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftClusterReconciler::get_common_labels(cluster)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            type_: Some("Opaque".to_string()),
            string_data: Some(BTreeMap::from([(
                SECRET_DATA_KEY.to_string(),
                Self::create_proxy_guard_key(),
            )])),
            ..Secret::default()
        };

        Ok(secret)
    }

    async fn update(
        &self,
        _cluster: &Self::OwnerType,
        _secret: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        Ok(())
    }
}

impl ForwardingSecretBuilder {
    pub fn new(client: Client) -> Self {
        ForwardingSecretBuilder { client }
    }

    fn create_proxy_guard_key() -> String {
        let mut rng = rand::thread_rng();
        Alphanumeric.sample_string(&mut rng, 64)
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER},
    };

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::ForwardingSecretBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-forwarding-secret");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ForwardingSecretBuilder::new(client);

        // W
        let secret = builder
            .create(&TEST_CLUSTER, "my-cluster-forwarding-secret")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(secret, {
            ".stringData.key" => "[forwarding secret random redacted]"
        });
    }

    #[tokio::test]
    async fn create_has_secret() {
        // G
        let client = create_client_mock();
        let builder = super::ForwardingSecretBuilder::new(client);

        // W
        let secret = builder
            .create(&TEST_CLUSTER, "my-cluster-forwarding-secret")
            .await
            .unwrap();

        // T
        assert!(secret
            .string_data
            .unwrap()
            .get(super::SECRET_DATA_KEY)
            .is_some());
    }

    #[test]
    fn create_proxy_guard_key_random() {
        // W
        let (secret_1, secret_2, secret_3) = (
            super::ForwardingSecretBuilder::create_proxy_guard_key(),
            super::ForwardingSecretBuilder::create_proxy_guard_key(),
            super::ForwardingSecretBuilder::create_proxy_guard_key(),
        );

        // T
        assert_ne!(secret_1, secret_2);
        assert_ne!(secret_2, secret_3);
        assert_ne!(secret_3, secret_1);
    }
}
