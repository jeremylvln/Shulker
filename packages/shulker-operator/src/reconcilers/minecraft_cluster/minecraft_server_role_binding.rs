use k8s_openapi::api::rbac::v1::RoleBinding;
use k8s_openapi::api::rbac::v1::RoleRef;
use k8s_openapi::api::rbac::v1::Subject;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use super::minecraft_server_role::MinecraftServerRoleBuilder;
use super::minecraft_server_service_account::MinecraftServerServiceAccountBuilder;
use super::MinecraftClusterReconciler;
use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;

pub struct MinecraftServerRoleBindingBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for MinecraftServerRoleBindingBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = RoleBinding;

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-server", cluster.name_any())
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    fn is_needed(&self, _cluster: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let role_binding = RoleBinding {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftClusterReconciler::get_common_labels(cluster)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            role_ref: RoleRef {
                api_group: "rbac.authorization.k8s.io".to_string(),
                kind: "Role".to_string(),
                name: MinecraftServerRoleBuilder::name(cluster),
            },
            ..RoleBinding::default()
        };

        Ok(role_binding)
    }

    async fn update(
        &self,
        cluster: &Self::OwnerType,
        role_binding: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        role_binding.subjects = Some(vec![Subject {
            kind: "ServiceAccount".to_string(),
            name: MinecraftServerServiceAccountBuilder::name(cluster),
            namespace: cluster.namespace(),
            ..Subject::default()
        }]);

        Ok(())
    }
}

impl MinecraftServerRoleBindingBuilder {
    pub fn new(client: Client) -> Self {
        MinecraftServerRoleBindingBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use k8s_openapi::api::rbac::v1::{RoleRef, Subject};

    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER},
    };

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::MinecraftServerRoleBindingBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-server");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBindingBuilder::new(client);

        // W
        let role_binding = builder
            .create(&TEST_CLUSTER, "my-cluster-server")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(role_binding);
    }

    #[tokio::test]
    async fn create_has_role() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBindingBuilder::new(client);

        // W
        let role_binding = builder
            .create(&TEST_CLUSTER, "my-cluster-server")
            .await
            .unwrap();

        // T
        assert_eq!(
            role_binding.role_ref,
            RoleRef {
                api_group: "rbac.authorization.k8s.io".to_string(),
                kind: "Role".to_string(),
                name: "my-cluster-server".to_string(),
            }
        );
    }

    #[tokio::test]
    async fn update_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBindingBuilder::new(client);
        let mut role_binding = builder
            .create(&TEST_CLUSTER, "my-cluster-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_CLUSTER, &mut role_binding)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(role_binding);
    }

    #[tokio::test]
    async fn update_has_subjects() {
        // G
        let client = create_client_mock();
        let builder = super::MinecraftServerRoleBindingBuilder::new(client);
        let mut role_binding = builder
            .create(&TEST_CLUSTER, "my-cluster-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_CLUSTER, &mut role_binding)
            .await
            .unwrap();

        // T
        assert_eq!(
            role_binding.subjects,
            Some(vec![Subject {
                kind: "ServiceAccount".to_string(),
                name: "my-cluster-server".to_string(),
                namespace: Some("default".to_string()),
                ..Subject::default()
            }])
        )
    }
}
