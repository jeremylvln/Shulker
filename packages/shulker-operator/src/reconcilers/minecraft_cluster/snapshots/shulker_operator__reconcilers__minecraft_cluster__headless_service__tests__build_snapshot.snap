---
source: packages/shulker-operator/src/reconcilers/minecraft_cluster/headless_service.rs
expression: service
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app.kubernetes.io/component: minecraft-server-headless
    app.kubernetes.io/instance: service-my-cluster
    app.kubernetes.io/managed-by: shulker-operator
    app.kubernetes.io/name: service
    app.kubernetes.io/part-of: cluster-my-cluster
    minecraftcluster.shulkermc.io/name: my-cluster
  name: my-cluster-cluster
  namespace: default
spec:
  clusterIP: None
  ports: []
  selector:
    minecraftcluster.shulkermc.io/name: my-cluster

